frontends/p4/typeChecking/typeUnification.cpp

```diff
@@ -303,64 +302,142 @@ bool TypeUnification::unify(const IR::Node* errorPosition,
+                            return false;
+                    }
+                }
             }
             return true;
         } else if (auto st = src->to<IR::Type_StructLike>()) {
             // There is another case, in which each field of the source is unifiable with the
             // corresponding field of the destination, e.g., a struct containing tuples.
-            if (strct->fields.size() != st->fields.size()) {
+            bool convenient = false;
+            if (auto unkn = src->to<IR::Type_UnknownStruct>()) {
+                convenient = unkn->fromDefaultInitializer;
+            }
+            if ((strct->fields.size() != st->fields.size()) && !convenient) {
                 if (reportErrors)
                     TypeInference::typeError("%1%: Number of fields %2% in initializer different "
                                              "than number of fields in structure %3%: %4% to %5%",
                                              errorPosition, st->fields.size(),
                                              strct->fields.size(), st, strct);
                 return false;
             }
-
-            for (const IR::StructField* f : strct->fields) {
-                auto stField = st->getField(f->name);
-                if (stField == nullptr) {
-                    TypeInference::typeError("%1%: No initializer for field %2%", errorPosition, f);
-                    return false;
+            if (!convenient) {
+                for (const IR::StructField* f : strct->fields) {
+                    auto stField = st->getField(f->name);
+                    if (stField == nullptr) {
+                        TypeInference::typeError("%1%: No initializer for field %2%",
+                                                 errorPosition, f);
+                        return false;
+                    }
+                    bool success = unify(errorPosition, f->type, stField->type, reportErrors);
+                    if (!success)
+                        return false;
+                }
+            } else {
+                for (const IR::StructField* f : st->fields) {
+                    auto strctField = strct->getField(f->name);
+                    if (strctField == nullptr) {
+                        TypeInference::typeError("%1%: No field named %2% in structure %3%",
+                                                 errorPosition, f->name, strct);
+                        return false;
+                    }
+                }
+                for (const IR::StructField* f : strct->fields) {
+                    if (f->type->is<IR::Type_Varbits>() || f->type->is<IR::Type_Stack>()
+                        || f->type->is<IR::Type_HeaderUnion>()) {
+                        TypeInference::typeError("%1%: Structs, headers and tuples "
+                                                 "containing %2% data type as a lief "
+                                                 "field cannot be zero initialized with "
```

Comment:
haven't looked very carefully, but there is a fair amount of code duplication, please see if you can refactor the common code in a function.